### PR TITLE
logcli: add support for bearer token authentication

### PR DESCRIPTION
**Pull request by dbluxo**
**What this PR does / why we need it**:

Add support for basic token authentication in cases where the loki-queriers/frontends are exposed through a reverse proxy doing token/bearer authentication.

Based on closed PR: https://github.com/grafana/loki/pull/2889 Thanks @gcotone

**Which issue(s) this PR fixes**:
Fixes https://github.com/grafana/loki/issues/395 (Partially - it does not implement oauth2)

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [ ] Tests updated

### CODE REVIEW

**Comment by dbluxo** [comment]
@achatterjee-grafana I saw that you reviewed the PR that this is based on. How can I request a review for this PR?

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added some review comments.

File: docs/sources/getting-started/logcli.md [hunk 1]
```diff
@@ -225,58 +253,70 @@ usage: logcli labels [<flags>] [<label>]
...
+      --username=""           Username for HTTP basic auth. Can also be set using LOKI_USERNAME env var.
+      --password=""           Password for HTTP basic auth. Can also be set using LOKI_PASSWORD env var.
+      --ca-cert=""            Path to the server Certificate Authority. Can also be set using LOKI_CA_CERT_PATH env var.
+      --tls-skip-verify       Server certificate TLS skip verify.
+      --cert=""               Path to the client certificate. Can also be set using LOKI_CLIENT_CERT_PATH env var.
+      --key=""                Path to the client certificate key. Can also be set using LOKI_CLIENT_KEY_PATH env var.
+      --org-id=""             adds X-Scope-OrgID to API requests for representing tenant ID. Useful for requesting tenant data when bypassing an auth gateway.
+      --bearer-token=""       adds the Authorization header Bearer to API requests for authentication purposes. Can also be set using LOKI_BEARER_TOKEN env var.
+      --bearer-token-file=""  adds the Authorization header Bearer to API requests for authentication purposes. Can also be set using LOKI_BEARER_TOKEN_FILE env var.
```
Comment (achatterjee-grafana) [inline docs/sources/getting-started/logcli.md:312]:
```suggestion
      --bearer-token-file=""  Add the Authorization header Bearer to API requests for authentication purposes. Can also be set using LOKI_BEARER_TOKEN_FILE env var.
```

File: docs/sources/getting-started/logcli.md [hunk 2]
```diff
@@ -225,58 +253,70 @@ usage: logcli labels [<flags>] [<label>]
...
-      --to=TO            Stop looking for logs at this absolute time (exclusive).
-      --match=MATCH ...  eg '{foo="bar",baz=~".*blip"}'
+                              Server address. Can also be set using LOKI_ADDR env var.
+      --username=""           Username for HTTP basic auth. Can also be set using LOKI_USERNAME env var.
+      --password=""           Password for HTTP basic auth. Can also be set using LOKI_PASSWORD env var.
+      --ca-cert=""            Path to the server Certificate Authority. Can also be set using LOKI_CA_CERT_PATH env var.
+      --tls-skip-verify       Server certificate TLS skip verify.
+      --cert=""               Path to the client certificate. Can also be set using LOKI_CLIENT_CERT_PATH env var.
+      --key=""                Path to the client certificate key. Can also be set using LOKI_CLIENT_KEY_PATH env var.
+      --org-id=""             adds X-Scope-OrgID to API requests for representing tenant ID. Useful for requesting tenant data when bypassing an auth gateway.
```
Comment (achatterjee-grafana) [inline docs/sources/getting-started/logcli.md:310]:
```suggestion
      --org-id=""             Add X-Scope-OrgID to API requests for representing tenant ID. Useful for requesting tenant data when bypassing an auth gateway.
```

File: docs/sources/getting-started/logcli.md [hunk 3]
```diff
@@ -225,58 +253,70 @@ usage: logcli labels [<flags>] [<label>]
...
+      --bearer-token=""       adds the Authorization header Bearer to API requests for authentication purposes. Can also be set using LOKI_BEARER_TOKEN env var.
+      --bearer-token-file=""  adds the Authorization header Bearer to API requests for authentication purposes. Can also be set using LOKI_BEARER_TOKEN_FILE env var.
+      --since=1h              Lookback window.
+      --from=FROM             Start looking for logs at this absolute time (inclusive)
+      --to=TO                 Stop looking for logs at this absolute time (exclusive)
+      --analyze-labels        Printout a summary of labels including count of label value combinations, useful for debugging high cardinality series
```
Comment (achatterjee-grafana) [inline docs/sources/getting-started/logcli.md:316]:
```suggestion
      --analyze-labels        Prints out a summary of labels including the count of label value combinations, useful for debugging high cardinality series
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Found a few more.

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Added some doc change requests.

File: docs/sources/getting-started/logcli.md [hunk 4]
```diff
@@ -155,6 +166,12 @@ Commands:
   series [<flags>] <matcher>
     Run series query.
 
+    The "series" command will take the provided label matcher and return all the log streams found in the time window.
```
Comment (achatterjee-grafana) [inline docs/sources/getting-started/logcli.md:169]:
```suggestion
    The "series" command takes the label matcher (provided) and returns all the log streams found in the time window.
```

File: docs/sources/getting-started/logcli.md [hunk 5]
```diff
@@ -168,53 +185,64 @@ The "query" command is useful for querying for logs. Logs can be returned in a f
 
 The output of the log can be specified with the "-o" flag, for example, "-o raw" for the raw output format.
 
-The "query" command will output extra information about the query and its results, such as the API URL, set of common labels, and
-set of excluded labels. This extra information can be suppressed with the --quiet flag.
+The "query" command will output extra information about the query and its results, such as the API URL, set of common labels, and set of excluded labels. This extra information can be suppressed with the --quiet flag.
```
Comment (achatterjee-grafana) [inline docs/sources/getting-started/logcli.md:188]:
```suggestion
The "query" command returns extra information about the query and its results, such as the API URL, a set of common labels, and a set of excluded labels. This extra information can be hidden with the --quiet flag.
```

**Review by achatterjee-grafana — CHANGES_REQUESTED** [review]
Some more.

File: docs/sources/getting-started/logcli.md [hunk 6]
```diff
@@ -93,24 +93,25 @@ usage: logcli [<flags>] <command> [<args> ...]
...
-      --stats            Show query statistics.
-  -o, --output=default   Specify output mode [default, raw, jsonl]. raw suppresses log labels and timestamp.
-  -z, --timezone=Local   Specify the timezone to use when formatting output timestamps [Local, UTC].
-      --cpuprofile=""    Specify the location for writing a CPU profile.
-      --memprofile=""    Specify the location for writing a memory profile.
+      --help                  Show context-sensitive help (also try --help-long and --help-man).
+      --version               Show application version.
+  -q, --quiet                 Suppress query metadata.
+      --stats                 Show query statistics.
+  -o, --output=default        Specify output mode [default, raw, jsonl]. raw suppresses log labels and timestamp.
```
[the remark this answers is not in the document]
Comment (dbluxo) [reply docs/sources/getting-started/logcli.md:100]: @achatterjee-grafana I don't understand the sense behind these improvements in the `getting-started/logcli.md`. This text is from the [source code](https://github.com/grafana/loki/blob/0a02aecdb993da9d56714d756a6be73039e5317d/cmd/logcli/main.go#L26), what is the point of changing it here?

**Comment by achatterjee-grafana** [comment]
Feel free to ignore the doc change req. I did not realize this was for Loki

**Comment by dbluxo** [comment]
What else can we do to move things forward here?

**Comment by dbluxo** [comment]
@slim-bean Maybe you could please add a review request from loki-team again, it somehow got kicked out and I don't have rights to add it :-(

**Comment by dbluxo** [comment]
There are conflicts in the meantime, I'm closing this PR and creating a new one with fewer doc changes as well.

**Pull request closed without merging**